minimal_memory_lock/minimal_memory_lock.cpp

```diff
@@ -0,0 +1,273 @@
+        this->publisher_->publish(message);
+        auto [minor, major] = get_new_page_faults();
+        RCLCPP_INFO(
+          this->get_logger(), "Page faults during spin: [minor: %d, major: %d]", minor, major);
+      };
+    timer_ = this->create_wall_timer(500ms, timer_callback);
+  }
+
+private:
+  rclcpp::TimerBase::SharedPtr timer_;
+  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
+  size_t count_;
+};
+
+int main(int argc, char * argv[])
+{
+  // Force flush of the stdout buffer.
+  setvbuf(stdout, NULL, _IONBF, BUFSIZ);
+
+  // Argument count and usage
+  if (rcutils_cli_option_exist(argv, argv + argc, "-h")) {
+    print_usage();
+    return 0;
+  }
+
+  // Configuration variables
+  size_t max_process_memory = 0;
+  size_t stack_size = 0;
+  bool lock_memory = false;
+  bool use_sleep = false;
+
+  // Optional argument parsing
+  if (rcutils_cli_option_exist(argv, argv + argc, OPTION_LOCK_MEMORY)) {
+    lock_memory = true;
+  }
+  if (rcutils_cli_option_exist(argv, argv + argc, OPTION_MEMORY_PREALLOCATION)) {
+    max_process_memory = 1024 * 1024 *
+      std::stoul(rcutils_cli_get_option(argv, argv + argc, OPTION_MEMORY_PREALLOCATION));
+  }
+  if (rcutils_cli_option_exist(argv, argv + argc, OPTION_THREAD_STACK_SIZE)) {
+    stack_size = 1024 *
+      std::stoul(rcutils_cli_get_option(argv, argv + argc, OPTION_THREAD_STACK_SIZE));
+  }
+  if (rcutils_cli_option_exist(argv, argv + argc, OPTION_USE_SLEEP)) {
+    use_sleep = true;
+  }
+
+  // Change a new default thread stack size to reduce memory foot-print
+  if (stack_size > 0) {
+    set_default_thread_stacksize(stack_size);
+  }
+
+  print_process_memory("Process memory before node creation: ");
+
+  // Create nodes
+  rclcpp::init(argc, argv);
+  auto node = std::make_shared<MinimalPublisher>();
+
+  // Sleep required to avoid some initial minor page faults
+  // TODO(carlos): find out why we see minor page faults initially
+  // This could be caused by middleware threads being created concurrently
+  if (use_sleep) {
+    std::cout << "Sleeping here" << std::endl;
+    std::this_thread::sleep_for(500ms);
+  }
+
+  if (lock_memory) {
+    print_process_memory("Process memory before locking: ");
+    bool memory_is_locked = configure_malloc_behavior();
+    if (memory_is_locked) {
+      std::cout << "Memory locked" << std::endl;
+    } else {
+      return EXIT_FAILURE;
+    }
+    print_process_memory("Process memory after locking: ");
+  }
+  if (max_process_memory > 0) {
+    std::cout << "Memory to reserve: " <<
+      std::to_string(max_process_memory / (1024 * 1024)) << " MB" << std::endl;
+    auto current_process_memory = get_process_memory() * 1024;
+    if (current_process_memory > max_process_memory) {
+      std::cout << "Process memory higher than maximum memory to reserve: " << std::endl;
+      return EXIT_FAILURE;
+    }
+    auto additional_memory_to_reserve = max_process_memory - current_process_memory;
```

Comment:
Is there a reason to do this complex operation here instead of just reserving a constant size block of memory? It seems like it should accomplish the same thing.

Reply:
Good point. That's how I implemented it at first. I considered these 2 options:

1. Reserve some additional memory. You create all the resources your application needs. If the application is truly static (no additional memory is required) then you don't have to reserve more memory. Actually, you don't even need to use `MCL_FUTURE` at all. If this is not the case, and the application may allocate more memory, you want to add some additional memory pool.
2. Set the maximum allowed reserved memory for the application: In this case, I don't want to reserve additional memory, just up the specified amount. This way you know exactly the memory your application is reserving. If for some reason, when creating all the resources in the initialization phase, this memory is exceeded, you get an error so you know that you will have to pre-allocate more memory for your application. For this example, if we set a maximum of 100MB, the total process memory is 100MB for Cyclone and for Fast-DDS. For Iceroyx, the application reserves more than 100MB and we get an error. Then we could reconsider the preallocated memory size and use 200MB.

An alternative approach would be to reserve the memory first and then create all the resources. However, you could be exceeding the reserved memory pool without knowing it. 

If you think option 1 is simpler I would use that option and rename the option so it's clear it is an **additional memory** pool.

Reply:
> Actually, you don't even need to use MCL_FUTURE at all.

IIRC, without `MCL_FUTURE`, thread stacks created after the mlockall call (mlockall traditionally is called in the main thread, before the creation of child threads) will not be automatically locked to RAM (as demand paging would apply again to those threads).

Reply:
Yes, but note I'm using a different approach. I'm calling `mlockall` once I have created all the threads. After this point, no thread should be created or additional memory allocated.

Reply:
Ah I misread what you did. Sorry about that and sounds good. 

Also for your option 1, I think if you want to limit your total process memory to some amount _X_, then your approach makes sense. I didn't know what was your intention. Perhaps documenting that intention in a sentence in the README might be helpful.